src/PureScript/CST/Parser.purs

```diff
@@ -97,9 +97,11 @@ layoutNonEmpty valueParser = ado
 
 layout :: forall a. Parser a -> Parser (Array a)
 layout valueParser =
-  tokLayoutStart *> values <* tokLayoutEnd
+  tokLayoutStart *> values
   where
-  values = (go =<< valueParser) <|> pure []
```

Comment:
This seems like it should be equivalent to `optional`? If we had an empty layout, then valueParser should immediately fail if given a TokLayoutEnd, shouldn't it? It's not obvious to me what "backtracking far enough" means then, and that there isn't some other internal bug. Do we know where backtracking is halting?

Reply:
Okay, I was a little hasty in ascribing it to backtracking. Putting `try` in there does not help.

What's really messing it up is the indentation-based parser recovery when we call it in `parseAdo` (this is the only usage of `layout`). An an example change, this manages to parse the empty `ado in`:

```diff
--- a/src/PureScript/CST/Parser.purs
+++ b/src/PureScript/CST/Parser.purs
@@ -673,7 +673,7 @@ parseDo = do
 parseAdo :: Parser (Recovered Expr)
 parseAdo = do
   keyword <- tokQualifiedKeyword "ado"
-  statements <- layout (recoverDoStatement parseDoStatement)
+  statements <- layout parseDoStatement
   in_ <- tokKeyword "in"
   result <- parseExpr
   pure $ ExprAdo { keyword, statements, in: in_, result }
```

Reply:
It makes sense to me that there is something odd with recovery, but I'm not sure what. In the case of `ado in`, I would expect `recoverIndent` to immediately hit a `TokLayoutEnd` where `col == indent`, resulting in empty `tokens`, yielding `Nothing`, which would propagate the error.
https://github.com/natefaubion/purescript-language-cst-parser/blob/5afff30ddafc84317d35a4d091a00563d818b899/src/PureScript/CST/Parser.purs#L1155-L1164
https://github.com/natefaubion/purescript-language-cst-parser/blob/5afff30ddafc84317d35a4d091a00563d818b899/src/PureScript/CST/Parser/Monad.purs#L149-L151
Maybe something is wrong with the state threading in `recover`.

Reply:
Okay, I'll push a test for this case and dig deeper.